src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs

```diff
@@ -0,0 +1,75 @@
+ */
+
+using System;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using AspNet.Security.OAuth.Extensions;
+using Microsoft.AspNet.Authentication;
+using Microsoft.AspNet.Authentication.OAuth;
+using Microsoft.AspNet.Http.Authentication;
+using Microsoft.AspNet.WebUtilities;
+using Microsoft.Extensions.Internal;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Instagram {
+    public class InstagramAuthenticationHandler : OAuthHandler<InstagramAuthenticationOptions> {
+        public InstagramAuthenticationHandler([NotNull] HttpClient client)
+            : base(client) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var userInformationEndpoint = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, name: "access_token", value: tokens.AccessToken);
+
+            if (Options.UseSignedRequests) {
+                var signature = SignRequest("/users/self",
+                    tokens.AccessToken, Options.ClientSecret);
+
+                userInformationEndpoint = QueryHelpers.AddQueryString(userInformationEndpoint, name: "sig",
+                    value: signature);
+            }
+
+            var request = new HttpRequestMessage(HttpMethod.Get, userInformationEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+            var response =
+                await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            response.EnsureSuccessStatusCode();
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync()).Value<JObject>("data");
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, InstagramAuthenticationHelper.GetIdentifier(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Name, InstagramAuthenticationHelper.GetFullName(payload), Options.ClaimsIssuer);
+
+            var context = new OAuthCreatingTicketContext(Context, Options, Backchannel, tokens, payload) {
+                Principal = new ClaimsPrincipal(identity),
+                Properties = properties
+            };
+
+            await Options.Events.CreatingTicket(context);
+
+            if (context.Principal?.Identity == null) {
+                return null;
+            }
+
+            return new AuthenticationTicket(context.Principal, context.Properties, context.Options.AuthenticationScheme);
+        }
+
+        private static string SignRequest([NotNull] string endpoint, [NotNull] string accessToken,
+            string secret) {
+            var token = Encoding.UTF8.GetBytes($"{endpoint}|access_token={accessToken}");
```

Comment:
I checked this, and there are some issues with it. 
* The QueryHelpers.ParseQuery doesn't work as I expected. It seems to parse the url up until the first parameter as the first parameter. So http://localhost:54540/signin?key=value will get parsed to {http://localhost:54540/signin?key, value}. Seems like a bug to me.
* The values of the parameters are all strings. Doing `from value in parameter.Value orderby value`will therefore result in adding a pair for each character in value. I think those two lines can be omitted. 

Finally a question for my curiosity :smile:, why make this method protected? Are other classes in this namespace going to use this method?